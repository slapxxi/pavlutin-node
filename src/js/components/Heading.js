import { kebabCase } from 'lodash';
import React, { Children } from 'react';
import Icon from 'react-fontawesome';

function Heading({ level, children }) {
  const Component = matchHeadingLevel(level);
  const id = generateId(children);
  if (id === '') {
    return <Component>{ children }</Component>;
  }
  return (
    <Component id={id} className="heading">
      { children }
      <a href={id} className="heading__anchor">
        {' '}
        <Icon name="link" />
      </a>
    </Component>
  );
}

function matchHeadingLevel(level) {
  switch (level) {
    case 1: return props => <h1 {...props} />;
    case 2: return props => <h2 {...props} />;
    case 3: return props => <h3 {...props} />;
    case 4: return props => <h4 {...props} />;
    case 5: return props => <h5 {...props} />;
    case 6: return props => <h6 {...props} />;
    default: return props => <h1 {...props} />;
  }
}

function generateId(children) {
  if (Children.count(children) === 1) {
    return `#${kebabCase(children)}`;
  }
  if (Children.count(children) > 1) {
    return `#${kebabCase(children[0])}`;
  }
  return '';
}

export default Heading;
