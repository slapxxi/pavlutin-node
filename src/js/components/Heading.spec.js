import React from 'react';
import { shallow } from 'enzyme';
import Heading from './Heading';

const levels = [1, 2, 3, 4, 5, 6];

levels.forEach((l) => {
  it(`renders heading with level ${l}`, () => {
    const wrapper = shallow(<Heading level={l} />);
    expect(wrapper.html()).toBe(`<h${l}></h${l}>`);
  });
});

it('generates heading id based on content', () => {
  const wrapper = shallow(<Heading level={1}>Test It</Heading>);
  expect(wrapper.prop('id')).toBe('#test-it');
});

it('generates link to generated id', () => {
  const wrapper = shallow(<Heading level={5}>Test It</Heading>);
  expect(wrapper.find('a').prop('href')).toBe('#test-it');
});
