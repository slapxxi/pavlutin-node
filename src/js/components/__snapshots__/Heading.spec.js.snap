// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders heading with level 1 1`] = `<h1 />`;

exports[`renders heading with level 2 1`] = `<h2 />`;

exports[`renders heading with level 3 1`] = `<h3 />`;

exports[`renders heading with level 4 1`] = `<h4 />`;

exports[`renders heading with level 5 1`] = `<h5 />`;

exports[`renders heading with level 6 1`] = `<h6 />`;
